File: CodingInterviewsPythonSolutions/First Unique Number.py
from collections import OrderedDict

class FirstUnique:

    def __init__(self, nums: List[int]):
        self.uniqueNumbers = []
        self.valueToNode = OrderedDict()
        for num in nums:
            self.add(num)

    def showFirstUnique(self) -> int:
        if self.uniqueNumbers:
            return self.uniqueNumbers[0]
        return -1

    def add(self, value: int) -> None:
        if value in self.valueToNode:
            if self.valueToNode[value] is not None:
                self.uniqueNumbers.remove(self.valueToNode[value])
                self.valueToNode[value] = None
        else:
            self.uniqueNumbers.append(value)
            self.valueToNode[value] = value

File: CodingInterviewsJavaSolutions/First Unique Number.java
class FirstUnique {
    private Map<Integer, Integer> valueToNode;
    private List<Integer> uniqueNumbers;

    public FirstUnique(int[] nums) {
        valueToNode = new LinkedHashMap<>();
        uniqueNumbers = new ArrayList<>();
        for (int num : nums) {
            add(num);
        }
    }

    public int showFirstUnique() {
        if (!uniqueNumbers.isEmpty()) {
            return uniqueNumbers.get(0);
        }
        return -1;
    }

    public void add(int value) {
        if (valueToNode.containsKey(value)) {
            if (valueToNode.get(value) != null) {
                uniqueNumbers.remove(valueToNode.get(value));
                valueToNode.put(value, null);
            }
        } else {
            uniqueNumbers.add(value);
            valueToNode.put(value, value);
        }
    }
}
